src/ApiCompat/Microsoft.DotNet.ApiCompatibility/Extensions/SymbolExtensions.cs

```diff
@@ -29,7 +29,10 @@ private static SymbolDisplayFormat GetSymbolDisplayFormat()
             return format.WithParameterOptions(format.ParameterOptions & ~SymbolDisplayParameterOptions.IncludeParamsRefOut);
         }
 
-        internal static string ToComparisonDisplayString(this ISymbol symbol) => symbol.ToDisplayString(Format);
+        internal static string ToComparisonDisplayString(this ISymbol symbol) =>
+            symbol.ToDisplayString(Format)
+                  .Replace("System.IntPtr", "nint") // Treat IntPtr and nint as the same
```

Comment:
Will this create confusing diagnostics when someone is actually using `IntPtr`?  Say I had an API that used `IntPtr` and I removed it on the right, what diagnostic do we emit?

Reply:
This code path doesn't impact the `Target` that is being emitted by rules and APICompat. The `ToComparisonDisplayString` helper function is only used by the mapping infrastructure to map members together.

I.e. https://github.com/dotnet/sdk/blob/0975222d0666c8a5a0240e07830a71a900e33b96/src/ApiCompat/Microsoft.DotNet.ApiCompatibility/Rules/AttributesMustMatch.cs#L73 doesn't use that function to for encoding the target.

Reply:
This string is only used for comparing symbols, so the diagnostic that is emitted in that case is still `"CP0002 : Member 'CompatTests.First.H(System.IntPtr)' exists on left but not on right"`.

Reply:
If we do end up having to normalize this in comparisons, it might be better to do here:
https://github.com/dotnet/sdk/blob/0975222d0666c8a5a0240e07830a71a900e33b96/src/ApiCompat/Microsoft.DotNet.ApiCompatibility/DefaultSymbolsEqualityComparer.cs#L14

And only after the equality check fails, and only if references are not provided.  I'd like us to first see if we can get roslyn to behave as if the feature is supported.

Reply:
IMO the `ToComparisonDisplayString` method is the right place as the name already suggests that it calculates a display string **for comparison**.

Reply:
I see.  That makes sense.  I was mistaken about where this was used.  Is there any concern about applying this transformation even when we have references?

Reply:
This is interesting. I also considered doing the replacement the other way but imagined that `Replace("nint", ...)` would have more false positives than `Replace("System.IntPtr", ...)` because of length of the string and the "System classification".

Are dots `.` allowed in many members types in C#/VB? I.e. a method comparison string `Companint` would be replaced to `CompaSystem.IntPtr`.

Reply:
> This still works. Is there a reason we might prefer one direction over the other?

`IntPtr` has been around forever so there will be a lot more uses of it that actually mean `IntPtr` and not `nint`.  `nint` is new and while it is important, I don't expect it would be used as much as `IntPtr`.  From the same perspective, if we can make this an optional normalization that's only applied when references are absent it could further reduce the amount of unnecessary work that is done.

Reply:
Testing this on .NET 472 and .NET 7,
```C#
if (RuntimeFeature.IsSupported("NumericIntPtr"))
{
    return symbol.ToDisplayString(Format);
}
return symbol.ToDisplayString(Format)
    .Replace("nint", "System.IntPtr") // Treat nint and IntPtr as the same
    .Replace("nuint", "System.UIntPtr"); // Treat nuint and UIntPtr as the same
```
tests fail on .NET 7. It seems like even when NumericIntPtr is defined, Roslyn returns different display strings.

Reply:
Off the top of my head the two concerning places are nested types (which probably here use a `+` instead of `.`) and explicit interface implementations (e.g. the full method name of `IEnumerable.GetEnumerator() => GetEnumerator();` I believe is `System.IEnumerable.GetEnumerator`.  So while "nint" => "System.IntPtr" probably wouldn't create any false positives, '.' probably isn't a guaranteed "eh, it's fine".

Reply:
Agreed that nint is fairly new but it's not unlikely that `nint` and `nuint` word parts are used in a member's name. I would expect that replacing "nint" with "System.IntPtr" would result be more false positives than the other way around. That said, these false positives shouldn't have any negative impact as far as I can tell.

Reply:
>if (RuntimeFeature.IsSupported("NumericIntPtr"))

This is mixing *tooling runtime* with target runtime.  The check would need to be against whatever the target references that were used when compiling the assembly.  Calling the feature `RuntimeFeature.IsSupported` in the tool runtime has no relevance since we might be using an older runtime to test assemblies built for a newer one. 

The way I was imagining this to work was we would add some parameter to `DefaultSymbolsEqualityComparer` like `bool honorRuntimeFeatures` and set that as `withReferences == true`.  Then we could say that when it is false we'd do the hack-up in `DefaultSymbolsEqualityComparer.GetKey` to normalize those symbols.

Reply:
Hmm, the premise of the `withReferences` suggestion was based on the fact that we only face this problem when we don't have the matching references.  However, we probably are in this state even when references are provided, since we don't have references for the left-hand-side and that might be the runtime that doesn't support the feature :(

I was really trying to get this to a place where we didn't need this hack indefinitely but I guess that might not be possible at the moment.